qiskit_nature/utils/random.py

```diff
@@ -0,0 +1,79 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Methods to sample random objects."""
+
+from typing import Any
+
+import numpy as np
+from qiskit.quantum_info import random_hermitian
+from qiskit_nature.operators.second_quantization import QuadraticHamiltonian
+
+
+# TODO see if type of seed can be specified instead of using Any
+def parse_random_seed(seed: Any) -> np.random.Generator:
+    """Parse a random number generator seed and return a Generator.
+
+    Args:
+        seed: The pseudorandom number generator or seed. Should be an
+            instance of `np.random.Generator` or else a valid input to
+            `np.random.default_rng`
+
+    Returns:
+        The np.random.Generator instance
+    """
+    if isinstance(seed, np.random.Generator):
+        return seed
+    return np.random.default_rng(seed)
+
+
+def random_antisymmetric_matrix(dim: int, seed: Any = None) -> np.ndarray:
+    """Return a random antisymmetric matrix.
+
+    Args:
+        dim: The width and height of the matrix.
+        seed: The pseudorandom number generator or seed. Should be an
+            instance of `np.random.Generator` or else a valid input to
+            `np.random.default_rng`
+
+    Returns:
+        The sampled antisymmetric matrix.
+    """
+    rng = parse_random_seed(seed)
+    mat = rng.standard_normal((dim, dim)) + 1j * rng.standard_normal((dim, dim))
+    return mat - mat.T
+
+
+def random_quadratic_hamiltonian(
```

Comment:
I do not see the need for such a specific method to be exposed here.

Reply:
This is why I originally named this file `testing.py` because this method is used for testing. But @woodsp-ibm asked to rename the file. @woodsp-ibm I've added another function here; are you okay with changing it back to `testing.py`?

Reply:
@mrossinek You mean that you don't think these functions should be exposed at the `qiskit.utils` level, right? In that case, I propose we create a new module `qiskit.utils.testing` and expose them there. What do you think? @woodsp-ibm thoughts?

Reply:
In Aqua we had random matrix generating functions and you could use them how you wanted whether that was in unit testing, experimenting with how a different algorithm behaved (which I guess is testing to some degree) or whatever. Question is do you name this/these for what they are or a what you think someone would do with them. A testing module could end up being a  hodge-podge of different things all used for "testing". Current I see its just used for unit testing - maybe rather than create this here for now just move it into the unit test or a support module there. If someone needs similar they can copy/use it from there and if we have enough demand it could be made part of public API rather than integrate it into the public API already.

Reply:
These functions are used in several different files throughtout the repo, so they should definitely live in a high-level location. Personally I use functions like these quite a lot so I would prefer them to be part of the public API, especially since I see no downside.

Reply:
>From what I can see the though parse_random_seed is still there. You expressed you wanted to be able to seed these independently. That I can accept, though we never did this before. My preference would be if seed is None that these run off the algorithm_globals random generator - similar to the quantum_info random ones you showed where they use a common generator if no seed is given - otherwise use some local one via the given seed

In [`qiskit.quantum_info.operators.random`](https://github.com/Qiskit/qiskit-terra/blob/89c4f6eb5c56f49ec5149e5be39c77d36b182b4e/qiskit/quantum_info/operators/random.py) the "global" genorator is simply `np.random.default_rng()`. My code also defers to `np.random.default_rng()` when the seed is None, so the behavior is equivalent. I simply factored out the function `parse_random_seed` because the function code was duplicated across the different random generation functions. As you requested, I removed it from the public API.

>In terms of (re-)moving them from here - they could still be there in test folder so I was not proposing getting rid of them entirely so you can still use them. As to part of public API is something we have to support/maintain them, document them, people translate docs etc versus just being there in unit test area so it extra effort/responsibility etc around these if we chose to have them there.

If they are in the test folder I can't import them from `qiskit_nature` which makes them inconvenient to use. I don't anticipate these functions causing a noticeable maintenance burden, and I don't mind taking responsibility for maintaining them.

Since you don't seem to be strongly opposed to the way I have things here, I ask that we please just keep it as it is.

Reply:
Ok, done. @mrossinek @woodsp-ibm are you all okay with how things are now?

Reply:
> If they are in the test folder I can't import them from qiskit_nature which makes them inconvenient to use.

Once the day comes that you may need to use them inside of `qiskit_nature`, we can promote them up. For now, since they are only being used in testing, they should go there.

However, my main point was that I do not see the need for the random `QuadraticHamiltonian`. This is a _very_ specific method. Once we have that, why don't we have methods for other operators? Having methods to generate random matrices is fine if you really have such a high demand for it, but having a wrapper purely around `QuadraticHamiltonian` I do not see the point of.

Reply:
>Once the day comes that you may need to use them inside of qiskit_nature, we can promote them up. For now, since they are only being used in testing, they should go there.

Most functions and classes exposed by Qiskit Nature are exposed because they are useful and not because they are used elsewhere in the library. There is already precedent in Qiskit Terra for exposing functions for generating random objects.

>Once we have that, why don't we have methods for other operators?

We should have such methods, they are actually very useful. Probably many existing tests could be simplified using them.

Anyway, I don't want to argue anymore so I've moved the functions out of `qiskit_nature`.